pyscf_ipu/experimental/nuclear_gradients.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+from typing import Callable
+import jax.numpy as jnp
+from jax import jit, tree_map, vmap
+from jax.ops import segment_sum
+
+from .basis import Basis
+from .integrals import _overlap_primitives, _kinetic_primitives
+from .orbital import batch_orbitals
+from .primitive import Primitive
+from .types import Float3, Float3xNxN
+
+
+def grad_primitive_integral(
+    primitive_op: Callable, atom_index: int, a: Primitive, b: Primitive
+) -> Float3:
+    """Generic gradient of a one-electron integral with respect the atom_index center"""
+    axes = jnp.arange(3)
+    lhs_p1 = vmap(a.offset_lmn, (0, None))(axes, 1)
+    t1 = 2 * a.alpha * vmap(primitive_op, (0, None))(lhs_p1, b)
+
+    lhs_m1 = vmap(a.offset_lmn, (0, None))(axes, -1)
+    t2 = jnp.where(a.lmn > 0, a.lmn, jnp.zeros_like(a.lmn))
+    t2 *= vmap(primitive_op, (0, None))(lhs_m1, b)
+    grad_out = t1 - t2
+    return jnp.where(a.atom_index == atom_index, grad_out, jnp.zeros_like(grad_out))
```

Comment:
This is a lot of computation to then get multiplied by zero -- can we not scatter/gather this from a higher level?   I.e. ensure this code is only called where atom_index = a.atom_index?